src/object/deduplicate_key.c

```diff
@@ -0,0 +1,497 @@
+    }
+
+  st_sm_atts_init = true;
+  return;
+
+error_exit:
+  if (att)
+    {
+      classobj_free_attribute (att);
+      assert_release (false);
+    }
+}
+
+SM_ATTRIBUTE *
+dk_find_sm_deduplicate_key_attribute (int att_id, const char *att_name)
+{
+  int level;
+
+  assert ((att_id != -1) || (att_name && *att_name));
+  if (att_id != -1)
+    {
+      assert (IS_DEDUPLICATE_KEY_ATTR_ID (att_id));
+      level = GET_DEDUPLICATE_KEY_ATTR_LEVEL (att_id);
+    }
+  else
+    {
+      GET_DEDUPLICATE_KEY_ATTR_MODE_LEVEL_FROM_NAME (att_name, level);
+    }
+
+  assert (level >= DEDUPLICATE_KEY_LEVEL_MIN && level <= DEDUPLICATE_KEY_LEVEL_MAX);
+  assert (dk_reserved_deduplicate_key_attr_name[level] != NULL);
+  assert (st_sm_atts_init == true);
+
+  return st_sm_atts[LEVEL_2_IDX (level)];
+}
+
+int
+dk_sm_deduplicate_key_position (int n_attrs, SM_ATTRIBUTE ** attrs, SM_FUNCTION_INFO * function_index)
+{
+  if (n_attrs > 1)
+    {
+      if (function_index)
+	{
+	  if ((function_index->attr_index_start > 0)
+	      && IS_DEDUPLICATE_KEY_ATTR_ID (attrs[function_index->attr_index_start - 1]->id))
+	    {
+	      return function_index->attr_index_start;
+	    }
+	}
+      else if (IS_DEDUPLICATE_KEY_ATTR_ID (attrs[n_attrs - 1]->id))
+	{
+	  return (n_attrs - 1);
+	}
+    }
+
+  return -1;
+}
+
+void
+dk_create_index_level_remove_adjust (DB_CONSTRAINT_TYPE ctype, char **attnames, int *asc_desc,
+				     int *attrs_prefix_length, SM_FUNCTION_INFO * func_index_info,
+				     int deduplicate_key_col_pos, int nnames)
+{
+  int func_no_args = 0;
+
+  assert (!DB_IS_CONSTRAINT_UNIQUE_FAMILY (ctype));
+  if (ctype != DB_CONSTRAINT_FOREIGN_KEY)
+    {
+      assert (asc_desc != NULL);
+    }
+
+  if (deduplicate_key_col_pos != -1)
+    {				// remove hidden column   
+      attnames[deduplicate_key_col_pos] = NULL;
+
+      assert (!func_index_info || (func_index_info && func_index_info->attr_index_start > 0));
+      if (func_index_info && func_index_info->attr_index_start > 0)
+	{
+	  func_no_args = nnames - deduplicate_key_col_pos;
+	  if (func_no_args > 0)
+	    {
+	      memmove (asc_desc + deduplicate_key_col_pos, asc_desc + (deduplicate_key_col_pos + 1),
+		       (func_no_args * sizeof (asc_desc[0])));
+	      if (attrs_prefix_length)
+		{
+		  memmove (attrs_prefix_length + deduplicate_key_col_pos,
+			   attrs_prefix_length + (deduplicate_key_col_pos + 1),
+			   (func_no_args * sizeof (attrs_prefix_length[0])));
+		}
+	      memmove (attnames + deduplicate_key_col_pos, attnames + (deduplicate_key_col_pos + 1),
+		       (func_no_args * sizeof (attnames[0])));
+
+	      attnames[nnames - 1] = NULL;
+	      func_index_info->attr_index_start--;
+	    }
+	}
+      deduplicate_key_col_pos = -1;
+    }
+}
+
+void
+dk_create_index_level_adjust (const PT_INDEX_INFO * idx_info, char **attnames, int *asc_desc,
+			      int *attrs_prefix_length, SM_FUNCTION_INFO * func_index_info, int nnames, bool is_reverse)
+{
+  int deduplicate_key_col_pos;
+
```

Comment:
doesn't need to check nnames value to have valid range?

Reply:
It is unlikely that the value of nnames is unusual when this function is called.
But let's add the assert() statement.